### PR TITLE
download_file should not create new folder

### PR DESCRIPTION
**Pull request by giorgiosironi**
It leads to unpleasant behaviour like
```
./bldr download_file:stackname,/tmp/file-123.tar.gz,file.tar.gz
// downloads to ./file.tar.gz/file-123.tar.gz
```
Instead we want to support the use cases shown in the test.

### CODE REVIEW

**Comment by lsh-0** [comment]
heh - of all the functions that frustrate me the most in builder, `download_file` and `upload_file` are *the* worst.

File: src/cfn.py [hunk 1]
```diff
@@ -233,7 +233,6 @@ def download_file(stackname, path, destination=None, allow_missing="False", use_
     """
     if not destination:
         destination = '.'
-    utils.mkdirp(destination)
```
Comment (lsh-0) [inline src/cfn.py]: we need a `ensure(os.path.exists(destination))` type call in there now

File: src/integration_tests/test_provisioning.py [hunk 2]
```diff
@@ -39,6 +40,10 @@ def test_create(self):
             lifecycle.start(stackname)
 
             cfn.cmd(stackname, "ls -l /", username=BOOTSTRAP_USER, concurrency='parallel')
+            cfn.download_file(stackname, "/bin/ls", "ls", use_bootstrap_user="true")
+            self.assertTrue(os.path.exists("./ls"))
```
Comment (lsh-0) [inline src/integration_tests/test_provisioning.py]: change to `os.path.isfile`

File: src/integration_tests/test_provisioning.py [hunk 3]
```diff
@@ -39,6 +40,10 @@ def test_create(self):
...
+            cfn.download_file(stackname, "/bin/less", "venv/bin/", use_bootstrap_user="true")
+            self.assertTrue(os.path.exists("./venv/bin/less"))
```
Comment (lsh-0) [inline src/integration_tests/test_provisioning.py]: add a test case to download to a destination that cannot be created that errors out

File: src/cfn.py (hunk 1 above)
Comment (giorgiosironi) [inline src/cfn.py]: but non existing targets are a supported use case like in `scp host:/home/elife/file.txt file.txt`

File: src/integration_tests/test_provisioning.py (hunk 2 above)
Comment (giorgiosironi) [inline src/integration_tests/test_provisioning.py]: yes, it could be stricter

File: src/integration_tests/test_provisioning.py (hunk 3 above)
Comment (giorgiosironi) [inline src/integration_tests/test_provisioning.py]: ok, should be catchable

File: src/integration_tests/test_provisioning.py [hunk 4]
```diff
@@ -39,6 +40,13 @@ def test_create(self):
             lifecycle.start(stackname)
 
             cfn.cmd(stackname, "ls -l /", username=BOOTSTRAP_USER, concurrency='parallel')
+            cfn.download_file(stackname, "/bin/ls", "ls", use_bootstrap_user="true")
+            self.assertTrue(os.path.isfile("./ls"))
+            cfn.download_file(stackname, "/bin/less", "venv/bin/", use_bootstrap_user="true")
+            self.assertTrue(os.path.isfile("./venv/bin/less"))
+
+            cfn.download_file(stackname, "/bin/pwd", "subfolder/pwd", use_bootstrap_user="true")
```
Comment (giorgiosironi) [inline src/integration_tests/test_provisioning.py]: I added this that I thought would fail, but it seems to work and to create `subfolder` first.

**Comment by giorgiosironi** [comment]
Back to @lsh-0 for additional review.

File: src/integration_tests/test_provisioning.py (hunk 4 above)
Comment (lsh-0) [inline src/integration_tests/test_provisioning.py]: I ran the tests locally and can see `venv/bin/less` , `./ls` and `subfolder/pwd` are still hanging around. I'll tweak the code to remove them after running the test.

**Comment by lsh-0** [comment]
if you're happy, I'm happy.

:+1:

**Comment by lsh-0** [comment]
huh, this is interesting:
```

03:48:53 Error in integration_tests.test_provisioning.TestProvisioning.test_create
03:48:53   File "/usr/lib/python2.7/unittest/case.py", line 358, in run
03:48:53     self.tearDown()
03:48:53   File "/ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP@2/src/integration_tests/test_provisioning.py", line 40, in tearDown
03:48:53     os.rmdir(path)
03:48:53 OSError: [Errno 39] Directory not empty: '/ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP@2/subfolder'
03:48:53 

...

03:48:53 [54.208.4.10] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP@2/ls <- /bin/ls
03:48:53 [54.208.4.10] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP@2/venv/bin/less <- /bin/less
03:48:53 [54.208.4.10] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP@2/subfolder/pwd <- /bin/pwd
03:48:53 should be empty: ['ls']
```

it's saying `./subfolder` cannot be deleted because it contains `ls`, however the only thing downloaded to `subfolder` is `pwd` ...

**Comment by lsh-0** [comment]
building again after manually removing `subfolder`. my guess is that it was probably leftover from a previous test

**Comment by lsh-0** [comment]
still failing ... checking code.

**Comment by giorgiosironi** [comment]
`ls` is consistently downloaded into `subfolder`:
```
elife@ci--alfred:/ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP$ ls -l subfolder/
total 108
-rw-r--r-- 1 jenkins jenkins 110080 Jun 20 11:34 ls
```

**Comment by giorgiosironi** [comment]
Output is:
```
12:44:05 [54.84.251.105] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP/ls <- /bin/ls
12:44:05 [54.84.251.105] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP/venv/bin/less <- /bin/less
12:44:05 [54.84.251.105] download: /ext/jenkins/workspace/er_PR-139-IRTU4SBFULEN3MXIRJAP/subfolder/pwd <- /bin/pwd
```
which looks correct

**Comment by giorgiosironi** [comment]
Ah, it was committed to the branch.

**Pull request merged**